stoneforge/pseudo_wells/monte_carlo_simulations.py

```diff
@@ -0,0 +1,310 @@
+
+  Parameters
+  ----------
+  data : array_like
+        1D dataset for which the experimental variogram must be calculated.
+  
+  rho : array_like
+        1D experimental correlation function of the data under examination.
+
+  Returns
+  -------
+  gama : array_like
+        1D experimental variogram of the data under examination.
+  
+  References
+  ----------
+  .. [1] Dvorkin, J.; Gutierrez, M. A.; Grana, D. Seismic reflections of rock
+  properties. [S.l.]: Cambridge University Press, 2014.
+
+  """
+  gama = (np.std(data)**2) * (1 - rho)
+  return(gama)
+
+# VARIOGRAM MODEL FIT
+def analytical_variogram(distance, gama):
+  """
+  Determines the best analytical variogram function to fit the experimental one, 
+  comparing the Gaussian, Exponential and Spherical models.
+
+  Parameters
+  ----------
+  distance : array_like
+        1D 
+  
+  gama : array_like
+        1D experimental variogram of the data under examination.
+
+  Returns
+  -------
+  model : array_like
+        1D analytical variogram that produces the best fit for the experimental variogram.
+  
+  References
+  ----------
+  .. [1] Dvorkin, J.; Gutierrez, M. A.; Grana, D. Seismic reflections of rock
+  properties. [S.l.]: Cambridge University Press, 2014.
+  .. [2] https://mmaelicke.github.io/scikit-gstat/reference/models.html
+  .. [3] https://mmaelicke.github.io/scikit-gstat/technical/fitting.html
+
+  """
+  # VARIOGRAM MODELS 
+  def sph(distance, range, sill):
+    return spherical(distance, range, sill)
+
+  def gauss(distance, range, sill):
+    return gaussian(distance, range, sill)
+
+  def exp(distance, range, sill):
+    return exponential(distance, range, sill)
+
+  # FIT
+  xi = distance
+  coeficients, cov = curve_fit(sph, distance, gama)                               
+  yi = list(map(lambda distancia: spherical(distancia, *coeficients), xi))        
+
+  xig = distance
+  coeficientsg, covg = curve_fit(gauss, distance, gama)
+  yig = list(map(lambda distancia: gaussian(distance, *coeficientsg), xig))
+
+  xie = distance
+  coeficientse, cove = curve_fit(exp, distance, gama)
+  yie = list(map(lambda distancia: exponential(distance, *coeficientse), xie))
+
+  # COMPARISON
+  ranges = np.array([coeficients[0],coeficientsg[0],coeficientse[0]])
+  structured_field = distance <= np.max(ranges)
+
+  difference_sph = np.zeros(len(gama))
+  difference_gauss = np.zeros(len(gama))
+  difference_exp = np.zeros(len(gama))
+
+  i = 0
+  while (structured_field[i] == True):
+    i += 1
+    difference_sph[i] = gama[i] - yi[i]
+    difference_gauss[i] = gama[i] - yig[i]
+    difference_exp[i] = gama[i] - yie[i]
+
+  rmse_sph = ((np.sum(difference_sph**2))/i)**0.5
+  rmse_gaus = ((np.sum(difference_gauss**2))/i)**0.5
+  rmse_exp = ((np.sum(difference_exp**2))/i)**0.5
+
+  RMSE = np.array((rmse_sph, rmse_gaus, rmse_exp))
+  best = list(RMSE).index(np.min(RMSE))
+
+  if best == 0:
+    return(yi)
+  if best == 1:
+    return(yig)
+  if best == 2:
+    return(yie)
```

Comment:
Gostei bastante da sua abordagem. Entretanto, mudaria um pouco. Acho que seria melhor deixar um parâmetro de entrada booleano default para esta função, que determinaria se faria o fit e usaria o melhor ajuste, ou se usaria um modelo de escolha do usuário de variograma.

Reply:
Oi Fábio, não sei exatamente como fazer isso usando uma entrada booleana, mas pensei em fazer desta forma pra que o usuário possa escolher um modelo ou usar o melhor ajuste: 

`# VARIOGRAM MODEL FIT
def analytical_variogram(distance: npt.ArrayLike, gama: npt.ArrayLike, model:str="best-fit")-> np.ndarray:
  """
  Fits the choosen analytical variogram function (model) to the experimental one, 
  if no model is choosen, determines the best model to fit, comparing the Gaussian, 
  Exponential and Spherical models.

  Parameters
  ----------
  distance : array_like
        1D array containing all the possible distances between a pair of points
        in the dataset.
  
  gama : array_like
        1D experimental variogram of the data under examination.
  
  model : str, optional
        Analytical variogram model to be fitted.  Should be one of:
            - "exponential": fits the exponential model
            - "gaussian": fits the gaussian model
            - "spherical": fits the spherical model
            - "best-fit": fits the three models above and verifies which one produces
              the smallest error.
        If not given, default method is "best-fit".

  Returns
  -------
  modeled_variogram : array_like
        The variogram model that has been choosen, or the variogram model that fits the
        best the experimental one. 
  
  References
  ----------
  .. [1] Dvorkin, J.; Gutierrez, M. A.; Grana, D. Seismic reflections of rock
  properties. [S.l.]: Cambridge University Press, 2014.
  .. [2] https://mmaelicke.github.io/scikit-gstat/reference/models.html
  .. [3] https://mmaelicke.github.io/scikit-gstat/technical/fitting.html

  """
  # VARIOGRAM MODELS 
  def sph(distance, range, sill):
    return spherical(distance, range, sill)

  def gauss(distance, range, sill):
    return gaussian(distance, range, sill)

  def exp(distance, range, sill):
    return exponential(distance, range, sill)

  if model == "spherical":
    xi = distance
    coeficients, cov = curve_fit(sph, distance, gama)                               
    yi = list(map(lambda distance: spherical(distance, *coeficients), xi))
    return(yi)

  elif model == "gaussian":
    xig = distance
    coeficientsg, covg = curve_fit(gauss, distance, gama)
    yig = list(map(lambda distance: gaussian(distance, *coeficientsg), xig))
    return(yig)

  elif model == "exponential":
    xie = distance
    coeficientse, cove = curve_fit(exp, distance, gama)
    yie = list(map(lambda distance: exponential(distance, *coeficientse), xie))
    return(yie)
  
  elif model == "best-fit":
    # FIT
    xi = distance
    coeficients, cov = curve_fit(sph, distance, gama)                               
    yi = list(map(lambda distance: spherical(distance, *coeficients), xi))        

    xig = distance
    coeficientsg, covg = curve_fit(gauss, distance, gama)
    yig = list(map(lambda distance: gaussian(distance, *coeficientsg), xig))

    xie = distance
    coeficientse, cove = curve_fit(exp, distance, gama)
    yie = list(map(lambda distance: exponential(distance, *coeficientse), xie))

    # COMPARISON
    ranges = np.array([coeficients[0],coeficientsg[0],coeficientse[0]])
    structured_field = distance <= np.max(ranges)

    difference_sph = np.zeros(len(gama))
    difference_gauss = np.zeros(len(gama))
    difference_exp = np.zeros(len(gama))

    i = 0
    while (structured_field[i] == True):
      i += 1
      difference_sph[i] = gama[i] - yi[i]
      difference_gauss[i] = gama[i] - yig[i]
      difference_exp[i] = gama[i] - yie[i]

    rmse_sph = ((np.sum(difference_sph**2))/i)**0.5
    rmse_gaus = ((np.sum(difference_gauss**2))/i)**0.5
    rmse_exp = ((np.sum(difference_exp**2))/i)**0.5

    RMSE = np.array((rmse_sph, rmse_gaus, rmse_exp))
    best = list(RMSE).index(np.min(RMSE))

    if best == 0:
      return(yi)
    if best == 1:
      return(yig)
    if best == 2:
      return(yie)
  else:
    raise TypeError("model must be: exponential, gaussian, spherical or best-fit")`